Thanks for the update - the caps, the merged dict, `format_file_size` and the new grouping test all look good. `format_file_size(size)` is the right call: with `BORG_UNITS` in place, `BORG_UNITS=raw` gives the exact byte count (`2160926 B`) for people who want it, `iec` gives MiB.

Some remaining points:

### Blocker: the two new truncation tests fail in binary mode

`check_cmd_test.py` is generated with `kinds="local,binary"`, and the `binary_archiver` fixture sets `EXE="borg.exe"` and `FORK_DEFAULT=True`, so borg runs in a *separate process*. `patch.object(ArchiveChecker, "MAX_MISSING_CHUNKS", 1)` in the test process has no effect there, the real caps apply, nothing gets truncated. Reproduced by putting a `borg.exe` wrapper on `PATH`:

```
FAILED check_cmd_test.py::test_missing_file_chunk_report_truncated[binary_archiver]
FAILED check_cmd_test.py::test_missing_file_chunk_refs_truncated[binary_archiver]
E  AssertionError: assert 'only the first 2 files are listed' in '...'
2 failed, 2 passed
```

The binary variants don't run in normal PR CI (only for tag builds and in the nightly canary workflow), which is why this is green here.

Two cheap fixes:

- `test_missing_file_chunk_refs_truncated` doesn't need patching at all: create 101 identical files and the real `MAX_REFS_PER_CHUNK = 100` fires. Verified against unpatched code - 100 reference lines, then `- ... (only the first 100 files are listed)`.
- `test_missing_file_chunk_report_truncated` can't do that (it would need 10001 missing chunks), so let it take the `archiver` fixture directly instead of `archivers` - the module's `pytest_generate_tests` only parametrizes when `archivers` is in the fixture names, so it then runs local-only.

### The report is lost on Ctrl-C or an exception

`report_missing_chunks()` runs only after the archive loop, and the per-chunk line is now `logger.debug`, so interrupting a long check of a badly damaged repo (or hitting the `Unknown archive metadata version` raise mid-loop) now prints nothing at all about missing chunks. Before, whatever had been found so far was already in the log.

Please wrap the loop in `try/finally`, and/or keep one error-level line on the *first* sighting of each chunk id - deduping by chunk id is exactly what removes the hardlink noise #9218 complained about, and it restores feedback during the run.

### The caps bound memory, but not to a small number

`MAX_MISSING_CHUNKS * MAX_REFS_PER_CHUNK` = up to ~1e6 reference entries. Measured with `tracemalloc` on that exact structure with ~90 char paths: ~385 MiB when each path is in 1-3 archives, ~2.3 GiB when it is in 24. It is also ~1e6 log lines, i.e. far past what anyone will read.

Since the grouped report is for humans, considerably lower caps (or an extra global cap on the total number of refs kept) would serve both memory and readability. The `--debug` lines are still there for the full picture.

### Truncation should say how much was dropped

`... (only the first 10000 missing chunks are listed)` doesn't tell whether 10001 or 10 million chunks are missing - probably the most interesting number in a damage report. A plain counter incremented in `record_missing_chunk()` is enough to print something like `10000 of N missing chunk references listed`.

### Nit

`missing_chunks[chunk_id] = [size, {}]` plus `entry[1]`: a tuple `(size, {})` unpacked as `size, refs = ...` reads better, `size` is never mutated.

--- a/src/borg/archive.py
+++ b/src/borg/archive.py
@@ -2138,19 +2138,21 @@ class ArchiveChecker:
         # Missing file chunks, collected during the per-archive checks and reported grouped as
         # chunk -> files -> archives after all archives were analyzed. Bounded by
         # MAX_MISSING_CHUNKS / MAX_REFS_PER_CHUNK.
-        missing_chunks = {}  # chunk_id -> [size, {path: {archive_name}}]
+        missing_chunks = {}  # chunk_id -> (size, {path: {archive_name}})
         missing_chunks_truncated = False  # True once the MAX_MISSING_CHUNKS cap was hit
         missing_refs_truncated = set()  # chunk_ids whose MAX_REFS_PER_CHUNK cap was hit
+        missing_refs_total = 0  # total missing chunk references seen (every file x chunk occurrence, uncapped)
 
         def record_missing_chunk(archive_name, path, chunk_id, size):
-            nonlocal missing_chunks_truncated
+            nonlocal missing_chunks_truncated, missing_refs_total
+            missing_refs_total += 1
             entry = missing_chunks.get(chunk_id)
             if entry is None:
                 if len(missing_chunks) >= self.MAX_MISSING_CHUNKS:
                     missing_chunks_truncated = True
                     return
-                entry = missing_chunks[chunk_id] = [size, {}]
-            refs = entry[1]
+                entry = missing_chunks[chunk_id] = (size, {})
+            size, refs = entry
             if path in refs:
                 refs[path].add(archive_name)
             elif len(refs) < self.MAX_REFS_PER_CHUNK:
@@ -2211,7 +2213,10 @@ class ArchiveChecker:
                 if chunk_id in missing_refs_truncated:
                     logger.error(f"    - ... (only the first {self.MAX_REFS_PER_CHUNK} files are listed)")
             if missing_chunks_truncated:
-                logger.error(f"... (only the first {self.MAX_MISSING_CHUNKS} missing chunks are listed)")
+                logger.error(
+                    f"... (only the first {self.MAX_MISSING_CHUNKS} missing chunks are listed; "
+                    f"{missing_refs_total} missing chunk references total)"
+                )
 
         def robust_iterator(archive):
             """Iterates through all archive items

--- a/src/borg/archive.py
+++ b/src/borg/archive.py
@@ -1851,8 +1851,8 @@ class RobustUnpacker:
 class ArchiveChecker:
     # Bound how many missing file chunks rebuild_archives buffers for its end-of-run report,
     # so checking a badly damaged repo with very many missing chunks can not exhaust memory.
-    MAX_MISSING_CHUNKS = 10000  # max. distinct missing chunk ids kept for the report
-    MAX_REFS_PER_CHUNK = 100  # max. referencing files kept per missing chunk
+    MAX_MISSING_CHUNKS = 1000  # max. distinct missing chunk ids kept for the grouped report
+    MAX_REFS_PER_CHUNK = 10  # max. referencing files kept per missing chunk
 
     def __init__(self):
         self.error_found = False
@@ -2152,6 +2152,8 @@ class ArchiveChecker:
                     missing_chunks_truncated = True
                     return
                 entry = missing_chunks[chunk_id] = (size, {})
+                # one line per chunk id (not per file), so an interrupted check still logs what it found.
+                logger.error(f"Missing chunk detected: {bin_to_hex(chunk_id)}, {format_file_size(size)}.")
             size, refs = entry
             if path in refs:
                 refs[path].add(archive_name)
@@ -2319,63 +2321,68 @@ class ArchiveChecker:
         pi = ProgressIndicatorPercent(
             total=num_archives, msg="Checking archives %3.1f%%", step=0.1, msgid="check.rebuild_archives"
         )
-        for i, info in enumerate(archive_infos):
-            pi.show(i)
-            archive_id, archive_id_hex = info.id, bin_to_hex(info.id)
-            try:
-                formatted = formatter.format_item(info, jsonline=False)
-            except (Archive.DoesNotExist, Repository.ObjectNotFound, IntegrityErrorBase):
-                # keys like {comment} need the archive metadata, which is damaged or missing here.
-                # use the values from the archive directory entry, they are always available.
-                formatted = f"{info.name} {OutputTimestamp(info.ts)} {archive_id_hex}"
-            logger.info(f"Analyzing archive {formatted} ({i + 1}/{num_archives})")
-            if archive_id not in self.chunks:
-                logger.error(f"Archive metadata block {archive_id_hex} is missing!")
-                self.error_found = True
+        # report the missing chunks collected so far even if the loop is interrupted (Ctrl-C) or aborts
+        # with an exception (e.g. the "Unknown archive metadata version" raise below), so a check of a
+        # badly damaged repo does not throw away everything it already found.
+        try:
+            for i, info in enumerate(archive_infos):
+                pi.show(i)
+                archive_id, archive_id_hex = info.id, bin_to_hex(info.id)
+                try:
+                    formatted = formatter.format_item(info, jsonline=False)
+                except (Archive.DoesNotExist, Repository.ObjectNotFound, IntegrityErrorBase):
+                    # keys like {comment} need the archive metadata, which is damaged or missing here.
+                    # use the values from the archive directory entry, they are always available.
+                    formatted = f"{info.name} {OutputTimestamp(info.ts)} {archive_id_hex}"
+                logger.info(f"Analyzing archive {formatted} ({i + 1}/{num_archives})")
+                if archive_id not in self.chunks:
+                    logger.error(f"Archive metadata block {archive_id_hex} is missing!")
+                    self.error_found = True
+                    if self.repair:
+                        logger.error(f"Deleting broken archive {info.name} {archive_id_hex}.")
+                        self.manifest.archives.delete_by_id(archive_id)
+                    else:
+                        logger.error(f"Would delete broken archive {info.name} {archive_id_hex}.")
+                    continue
+                cdata = self.repository.get(archive_id)
+                try:
+                    _, data = self.repo_objs.parse(archive_id, cdata, ro_type=ROBJ_ARCHIVE_META)
+                except IntegrityErrorBase as integrity_error:
+                    logger.error(f"Archive metadata block {archive_id_hex} is corrupted: {integrity_error}")
+                    self.error_found = True
+                    if self.repair:
+                        logger.error(f"Deleting broken archive {info.name} {archive_id_hex}.")
+                        self.manifest.archives.delete_by_id(archive_id)
+                    else:
+                        logger.error(f"Would delete broken archive {info.name} {archive_id_hex}.")
+                    continue
+                archive = self.key.unpack_archive(data)
+                archive = ArchiveItem(internal_dict=archive)
+                if archive.version != 2:
+                    raise Exception("Unknown archive metadata version")
+                items_buffer = ChunkBuffer(self.key)
+                items_buffer.write_chunk = add_callback
+                for item in robust_iterator(archive):
+                    if "chunks" in item:
+                        verify_file_chunks(info.name, item)
+                    items_buffer.add(item)
+                items_buffer.flush(flush=True)
                 if self.repair:
-                    logger.error(f"Deleting broken archive {info.name} {archive_id_hex}.")
-                    self.manifest.archives.delete_by_id(archive_id)
-                else:
-                    logger.error(f"Would delete broken archive {info.name} {archive_id_hex}.")
-                continue
-            cdata = self.repository.get(archive_id)
-            try:
-                _, data = self.repo_objs.parse(archive_id, cdata, ro_type=ROBJ_ARCHIVE_META)
-            except IntegrityErrorBase as integrity_error:
-                logger.error(f"Archive metadata block {archive_id_hex} is corrupted: {integrity_error}")
-                self.error_found = True
-                if self.repair:
-                    logger.error(f"Deleting broken archive {info.name} {archive_id_hex}.")
-                    self.manifest.archives.delete_by_id(archive_id)
-                else:
-                    logger.error(f"Would delete broken archive {info.name} {archive_id_hex}.")
-                continue
-            archive = self.key.unpack_archive(data)
-            archive = ArchiveItem(internal_dict=archive)
-            if archive.version != 2:
-                raise Exception("Unknown archive metadata version")
-            items_buffer = ChunkBuffer(self.key)
-            items_buffer.write_chunk = add_callback
-            for item in robust_iterator(archive):
-                if "chunks" in item:
-                    verify_file_chunks(info.name, item)
-                items_buffer.add(item)
-            items_buffer.flush(flush=True)
-            if self.repair:
-                archive.item_ptrs = archive_put_items(
-                    items_buffer.chunks, repo_objs=self.repo_objs, add_reference=add_reference
-                )
-                data = self.key.pack_metadata(archive.as_dict())
-                new_archive_id = self.key.id_hash(data)
-                logger.debug(f"archive id old: {bin_to_hex(archive_id)}")
-                logger.debug(f"archive id new: {bin_to_hex(new_archive_id)}")
-                cdata = self.repo_objs.format(new_archive_id, {}, data, ro_type=ROBJ_ARCHIVE_META)
-                add_reference(new_archive_id, len(data), cdata)
-                self.manifest.archives.create(info.name, new_archive_id, info.ts)
-                if archive_id != new_archive_id:
-                    self.manifest.archives.delete_by_id(archive_id)
-        pi.finish()
-        report_missing_chunks()
+                    archive.item_ptrs = archive_put_items(
+                        items_buffer.chunks, repo_objs=self.repo_objs, add_reference=add_reference
+                    )
+                    data = self.key.pack_metadata(archive.as_dict())
+                    new_archive_id = self.key.id_hash(data)
+                    logger.debug(f"archive id old: {bin_to_hex(archive_id)}")
+                    logger.debug(f"archive id new: {bin_to_hex(new_archive_id)}")
+                    cdata = self.repo_objs.format(new_archive_id, {}, data, ro_type=ROBJ_ARCHIVE_META)
+                    add_reference(new_archive_id, len(data), cdata)
+                    self.manifest.archives.create(info.name, new_archive_id, info.ts)
+                    if archive_id != new_archive_id:
+                        self.manifest.archives.delete_by_id(archive_id)
+        finally:
+            pi.finish()
+            report_missing_chunks()
 
     def finish(self):
         if self.repair:

--- a/src/borg/testsuite/archiver/check_cmd_test.py
+++ b/src/borg/testsuite/archiver/check_cmd_test.py
@@ -212,8 +212,9 @@ def test_missing_file_chunk(archivers, request):
     assert "The following chunks are missing in the repository:" not in output
 
 
-def test_missing_file_chunk_report_truncated(archivers, request):
-    archiver = request.getfixturevalue(archivers)
+def test_missing_file_chunk_report_truncated(archiver):
+    # local-only: this patches ArchiveChecker.MAX_MISSING_CHUNKS in-process, which has no effect
+    # when borg runs as a separate process (binary_archiver), so it must not be parametrized.
     check_cmd_setup(archiver)
 
     # remove several distinct file chunks, so more missing chunks exist than the (patched) report limit.
@@ -243,9 +244,11 @@ def test_missing_file_chunk_refs_truncated(archivers, request):
     archiver = request.getfixturevalue(archivers)
     cmd(archiver, "repo-create", RK_ENCRYPTION)
 
-    # several distinct files with identical content dedup to the same chunk, so a single missing
-    # chunk ends up referenced by many files, which exercises the per-chunk reference cap.
-    for i in range(5):
+    # many distinct files with identical content dedup to the same chunk, so a single missing chunk
+    # ends up referenced by more files than MAX_REFS_PER_CHUNK, which exercises the per-chunk cap
+    # without patching (so it works in binary mode too, where borg runs as a separate process).
+    cap = ArchiveChecker.MAX_REFS_PER_CHUNK
+    for i in range(cap + 1):
         create_regular_file(archiver.input_path, f"samefile{i}", contents=b"same content for dedup")
     cmd(archiver, "create", "archive1", "input")
 
@@ -259,12 +262,10 @@ def test_missing_file_chunk_refs_truncated(archivers, request):
                 break
     assert killed_id is not None
 
-    # cap references per chunk to 2, so the remaining referencing files are truncated.
-    with patch.object(ArchiveChecker, "MAX_REFS_PER_CHUNK", 2):
-        output = cmd(archiver, "check", exit_code=1)
+    output = cmd(archiver, "check", exit_code=1)
     assert "The following chunks are missing in the repository:" in output
     assert bin_to_hex(killed_id) in output
-    assert "only the first 2 files are listed" in output  # the remaining referencing files are truncated
+    assert f"only the first {cap} files are listed" in output  # the remaining referencing files are truncated
 
 
 def test_missing_archive_item_chunk(archivers, request):

--- a/src/borg/testsuite/archiver/check_cmd_test.py
+++ b/src/borg/testsuite/archiver/check_cmd_test.py
@@ -180,9 +180,12 @@ def test_missing_file_chunk(archivers, request):
 
     output = cmd(archiver, "check", exit_code=1)
     assert "The following chunks are missing in the repository:" in output
-    # archive1 and archive2 share src_file, so the missing chunk appears once, with both archives
-    # listed on its single reference line.
-    assert output.count(bin_to_hex(killed_chunk.id)) == 1
+    # archive1 and archive2 share src_file, so the missing chunk is grouped once, with both archives
+    # listed on its single reference line (the id also appears once in the streamed "Missing chunk
+    # detected" line emitted while the archives are analyzed).
+    killed_hex = bin_to_hex(killed_chunk.id)
+    chunk_header_lines = [ln for ln in output.splitlines() if ln.startswith("- Chunk ") and killed_hex in ln]
+    assert len(chunk_header_lines) == 1
     ref_lines = [line for line in output.splitlines() if src_file in line]
     assert len(ref_lines) == 1
     assert "archive1" in ref_lines[0] and "archive2" in ref_lines[0]

